### ISSUE TITLE
Feature Request: supporting ALOS as additional download source

### ISSUE THREAD

**Issue by keenonkites** [agrenott/pyhgtmap#41]
First of all Merry Christmas.

In order to have proper worldwide coverage for 1 asec tiles (also very north and very south) it would be great to have direct support for ALOS data source.

It looks that the download could be integrated quite easily as after the user registration you just have to know the path of the needed tile. The hierarchy for a 1 asec tile is something like 
`.../version>/<Folder_of_5x5_degrees>/<tile>`

You could also download a complete collection of 5x5 degrees (25 tiles in different version) as a zip, but I think that's not really helpful for pyhgtmap.

Cheers
Patrik

ALOS Global Digital Surface Model "ALOS World 3D - 30m (AW3D30)"
------------------------------------------------------------------------------
Below the details about data source and example download links:

Entry point with instructions (including user registration):
https://www.eorc.jaxa.jp/ALOS/en/dataset/aw3d30/aw3d30_e.htm

Acess to map for tile download:
https://www.eorc.jaxa.jp/ALOS/en/aw3d30/data/index.htm

direct download link for a single 1asec tile:
https://www.eorc.jaxa.jp/ALOS/aw3d30/data/release_v2303/N045E005/N046E007.zip

direct download link for 5 asec collection of 1asec tiles
https://www.eorc.jaxa.jp/ALOS/aw3d30/data/release_v2303/N045E005_N050E010.zip

**Comment by agrenott** [comment]
Hi and Merry Christmas :).
Looks feasible without too much hassle. I'll need to add user/password management to the plugin framework (can be shared with SRTM).

**Comment by keenonkites** [comment]
thanks... for both the good wishes and the good news regarding the enhancement.

Cheers
Patrik

**Comment by agrenott** [comment]
I actually started few weeks back, before getting sucked into something else. I'll resume WTP.

**Comment by keenonkites** [comment]
Cool, Thanks for implementing.

I still have no python 3.9 on the machine I will run pyhgtmap mainly. Therefor I can't actually test on the final machine due to that. We're trying to install Python 3.11 there, but I can't tell you when this is done.

Actually I'm trying it on WLS on my Windows 11 machine.... but I'm not sure if I get that running at all.

Keep you posted on any success.
Cheers
Patrik

**Comment by keenonkites** [comment]
Understandable.
Thanks for the advice. The update to higher version should have taken place earlier but couldn't be performed due to illness. Re-triggered it.

I got SRTM1 download already working on Windows 11 WSL, missing the gdal part right now but hope be able to perform this today in the evening.
Else I assume I can test the 'download' from ALOS also by using --download-only as a first try.

Cheers
Patrik

**Comment by keenonkites** [comment]
Update: I was able to get it running on Windows 11 WSL (Ubuntu).
ALOS1 as Datasource worked and generated outputfile. I was not able to test the generated data yet, but I do not assume we have a problem there. I'll probably will do that next few days.

I noticed that the download is extremely slow for ALOS1, but that has definitely nothing to do with pyhgtmap as my former tests with manual downloads showed the same behaviour.

Again thanks for implementing it, we now have a good quality source covering the completete world, not like SRTM only up to 60° (or so).
Cheers
Patrik

**Comment by keenonkites** [comment]
I have now python 3.12 on our Ubuntu 20.04.6 (installed from Deadsnakes PPA) and I was able to install pyhgtmap directly from git (after installing g++ for pybind11-rdp).
Unfortunately it fails here on Ubuntu with the following error message:
```
Traceback (most recent call last):
  File "/home/pab/venv_pyhgtmap3.12/bin/pyhgtmap", line 5, in <module>
    from pyhgtmap.main import main
  File "/home/pab/venv_pyhgtmap3.12/lib/python3.12/site-packages/pyhgtmap/main.py", line 10, in <module>
    from pyhgtmap import NASASRTMUtil
  File "/home/pab/venv_pyhgtmap3.12/lib/python3.12/site-packages/pyhgtmap/NASASRTMUtil.py", line 13, in <module>
    from matplotlib.path import Path as PolygonPath
  File "/home/pab/venv_pyhgtmap3.12/lib/python3.12/site-packages/matplotlib/__init__.py", line 161, in <module>
    from . import _api, _version, cbook, _docstring, rcsetup
  File "/home/pab/venv_pyhgtmap3.12/lib/python3.12/site-packages/matplotlib/rcsetup.py", line 32, in <module>
    from cycler import Cycler, cycler as ccycler
  File "/usr/lib/python3/dist-packages/cycler.py", line 48, in <module>
    from six.moves import zip, reduce
ModuleNotFoundError: No module named 'six.moves'
```

Can you tell me what I'm doing wrong or what I miss ? sorry, I'm not a specialist in python development.
As mentioned above, I did not have that on Windows running python3.11 ... is that difference of version (3.11 on Windows against 3.12 on Ubuntu) eventually the problem ?

Cheers
Patrik

**Comment by agrenott** [comment]
Looks like the "cycler" module is not properly installed in your 3.12 virtualenv. Python thus fallbacks to system's version which is probably much older and tries to import six, which is not available.
```
  File "**/home/pab/venv_pyhgtmap3.12/lib/python3.12/site-packages**/matplotlib/rcsetup.py", line 32, in <module>
    from cycler import Cycler, cycler as ccycler
  File "**/usr/lib/python3/dist-packages/cycler.py**", line 48, in <module>
    from six.moves import zip, reduce
```

Maybe try to re-install/upgrade cycler from inside your venv `pip install -U cycler`

**Comment by keenonkites** [comment]
cycler seems to be installed properly in the I think actual version 0.12.1 (it can be imported also in python):
```
$ pip show cycler
Name: cycler
Version: 0.12.1
Summary: Composable style cycles
Home-page: 
Author: 
Author-email: Thomas A Caswell <matplotlib-users@python.org>
License: Copyright (c) 2015, matplotlib project
All rights reserved.
...
Location: /home/pab/venv_pyhgtmap3.12/lib/python3.12/site-packages
Requires: 
Required-by: matplotlib
```

six would be installed also:
```
$ pip show six
Name: six
Version: 1.14.0
Summary: Python 2 and 3 compatibility utilities
Home-page: https://github.com/benjaminp/six
Author: Benjamin Peterson
Author-email: benjamin@python.org
License: MIT
Location: /usr/lib/python3/dist-packages
Requires: 
Required-by: oauth2client
```

I'm sort of lost....

**Comment by agrenott** [comment]
That's strange indeed, as your previous case obviously didn't use the venv
version of cycler.
The python 2 support (and thus dependency on six) has been removed in
cycler 0.11 (https://github.com/matplotlib/cycler/releases/tag/v0.11.0).

Le lun. 11 mars 2024 à 11:52, keenonkites ***@***.***> a
écrit :

**Comment by keenonkites** [comment]
I got the error messages when I create the 3.12 venv with the option --system-site-packages to have access to gdal.

```
Traceback (most recent call last):
  File "/home/pab/venv_pyhgtmap3.12-no-site-pkg/bin/pyhgtmap", line 5, in <module>
    from pyhgtmap.main import main
  File "/home/pab/venv_pyhgtmap3.12-no-site-pkg/lib/python3.12/site-packages/pyhgtmap/main.py", line 11, in <module>
    from pyhgtmap.cli import parse_command_line
  File "/home/pab/venv_pyhgtmap3.12-no-site-pkg/lib/python3.12/site-packages/pyhgtmap/cli.py", line 11, in <module>
    from pyhgtmap.hgt.file import parse_polygons_file
  File "/home/pab/venv_pyhgtmap3.12-no-site-pkg/lib/python3.12/site-packages/pyhgtmap/hgt/file.py", line 18, in <module>
    from .tile import HgtTile
  File "/home/pab/venv_pyhgtmap3.12-no-site-pkg/lib/python3.12/site-packages/pyhgtmap/hgt/tile.py", line 12, in <module>
    from pyhgtmap.hgt.contour import ContoursGenerator, build_contours
  File "/home/pab/venv_pyhgtmap3.12-no-site-pkg/lib/python3.12/site-packages/pyhgtmap/hgt/contour.py", line 8, in <module>
    from pybind11_rdp import rdp
  File "/home/pab/venv_pyhgtmap3.12-no-site-pkg/lib/python3.12/site-packages/pybind11_rdp/__init__.py", line 4, in <module>
    from _pybind11_rdp import LineSegment  # noqa
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ImportError: /home/pab/venv_pyhgtmap3.12-no-site-pkg/lib/python3.12/site-packages/_pybind11_rdp.cpython-312-x86_64-linux-gnu.so: undefined symbol: _ZTVN10__cxxabiv117__class_type_infoE
```

When creating it without --system-site-packages I get a different error message, not sure if I'm here already 'over' the other problem regarding cycler and six or not:

```
Traceback (most recent call last):
  File "/home/pab/venv_pyhgtmap3.12-no-site-pkg/bin/pyhgtmap", line 5, in <module>
    from pyhgtmap.main import main
  File "/home/pab/venv_pyhgtmap3.12-no-site-pkg/lib/python3.12/site-packages/pyhgtmap/main.py", line 11, in <module>
    from pyhgtmap.cli import parse_command_line
  File "/home/pab/venv_pyhgtmap3.12-no-site-pkg/lib/python3.12/site-packages/pyhgtmap/cli.py", line 11, in <module>
    from pyhgtmap.hgt.file import parse_polygons_file
  File "/home/pab/venv_pyhgtmap3.12-no-site-pkg/lib/python3.12/site-packages/pyhgtmap/hgt/file.py", line 18, in <module>
    from .tile import HgtTile
  File "/home/pab/venv_pyhgtmap3.12-no-site-pkg/lib/python3.12/site-packages/pyhgtmap/hgt/tile.py", line 12, in <module>
    from pyhgtmap.hgt.contour import ContoursGenerator, build_contours
  File "/home/pab/venv_pyhgtmap3.12-no-site-pkg/lib/python3.12/site-packages/pyhgtmap/hgt/contour.py", line 8, in <module>
    from pybind11_rdp import rdp
  File "/home/pab/venv_pyhgtmap3.12-no-site-pkg/lib/python3.12/site-packages/pybind11_rdp/__init__.py", line 4, in <module>
    from _pybind11_rdp import LineSegment  # noqa
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ImportError: /home/pab/venv_pyhgtmap3.12-no-site-pkg/lib/python3.12/site-packages/_pybind11_rdp.cpython-312-x86_64-linux-gnu.so: undefined symbol: _ZTVN10__cxxabiv117__class_type_infoE
```

If I then try to install gdal I can't:

```
      During handling of the above exception, another exception occurred:
      
      Traceback (most recent call last):
        File "<string>", line 190, in get_gdal_config
        File "<string>", line 92, in fetch_config
      gdal_config_error: [Errno 2] No such file or directory: 'gdal-config'
      
      Could not find gdal-config. Make sure you have installed the GDAL native library and development headers.
```

gdal is installed on the system but it correct, there's no gdal-config.

And right now I can't play around on the system too much (except in python) as the system is in heavy use right now.

I'm thinking about setting up a vm with same 'old' Ubuntu and try to get it running there, just to have something to play around I can't break.

**Comment by keenonkites** [comment]
Finally I was able to pick up this issue again (after blocked/busy server and absences).

I got it running now, not with normal venv but with miniconda environment.

Possible problem areas:

- system gdal installation old
- python-gdal not usable for my python 3.10/3.11/3.12 installation from deadsnakes (for Ubuntu 20.04
- pip install gdal not working: actual version requires newer version of system gdal, which I don't want/can't upgrade
- incorrect cached wheel file (from older attempts) for pybind11-rdp

miniconda environment installs much newer version of gdal with all needed tools. And after installing pybind11-rdp without cache everything seems to work, I'm able to build elevation files from alos1 sources.

Thanks again for your help regarding this.... and sorry for the long delay.

Cheers Patrik
